core/primitives/src/signature.rs

```diff
@@ -1,24 +1,121 @@
 extern crate exonum_sodiumoxide as sodiumoxide;
 
+use bs58;
+use std::fmt;
+
 pub use signature::sodiumoxide::crypto::sign::ed25519::Seed;
 
-pub type PublicKey = sodiumoxide::crypto::sign::ed25519::PublicKey;
-pub type SecretKey = sodiumoxide::crypto::sign::ed25519::SecretKey;
+#[derive(Copy, Clone, Eq, PartialOrd, Ord, PartialEq, Serialize, Deserialize)]
+pub struct PublicKey(pub sodiumoxide::crypto::sign::ed25519::PublicKey);
+
+#[derive(Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct SecretKey(pub sodiumoxide::crypto::sign::ed25519::SecretKey);
+
 pub type Signature = sodiumoxide::crypto::sign::ed25519::Signature;
 
 pub fn sign(data: &[u8], secret_key: &SecretKey) -> Signature {
-    sodiumoxide::crypto::sign::ed25519::sign_detached(data, secret_key)
+    sodiumoxide::crypto::sign::ed25519::sign_detached(data, &secret_key.0)
 }
 
 pub fn get_keypair_from_seed(seed: &Seed) -> (PublicKey, SecretKey) {
-    sodiumoxide::crypto::sign::ed25519::keypair_from_seed(seed)
+    let (public_key, secret_key) = sodiumoxide::crypto::sign::ed25519::keypair_from_seed(seed);
+    (PublicKey(public_key), SecretKey(secret_key))
 }
 
 pub fn get_keypair() -> (PublicKey, SecretKey) {
-    sodiumoxide::crypto::sign::ed25519::gen_keypair()
+    let (public_key, secret_key) = sodiumoxide::crypto::sign::ed25519::gen_keypair();
+    (PublicKey(public_key), SecretKey(secret_key))
 }
 
 pub fn default_signature() -> Signature {
     let sig = [0u8; sodiumoxide::crypto::sign::ed25519::SIGNATUREBYTES];
     sodiumoxide::crypto::sign::ed25519::Signature(sig)
 }
+
+impl PublicKey {
+    pub fn from(s: &str) -> PublicKey {
+        let mut array = [0; sodiumoxide::crypto::sign::ed25519::PUBLICKEYBYTES];
+        let bytes = bs58::decode(s).into_vec().expect("Failed to convert public key from base58");
+        assert_eq!(bytes.len(), array.len(), "decoded {} is not long enough for public key", s);
+        let bytes_arr = &bytes[..array.len()];
+        array.copy_from_slice(bytes_arr);
+        let public_key = sodiumoxide::crypto::sign::ed25519::PublicKey(array);
+        PublicKey(public_key)
+    }
+}
+
+impl<'a> From<&'a PublicKey> for String {
+    fn from(h: &'a PublicKey) -> Self {
+        bs58::encode(h.0).into_string()
```

Comment:
There are two important things other coins do with their base58 encoded strings:
1. checksum, to catch transcription errors
2. prepend with some sort of unique identifier, in order to differentiate from addresses from other coins (even testnets typically use a different identifier than the mainnet).

Although base58 is pretty ubiquitous, it's falling out of style in favor of things like [bech32](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki) and [eip55](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md)

Reply:
Well, in our case PublicKey is not an address, the AccountId is an address. So first we should actually figure out if we want AccountIds to plain text or hash of it.
I'm ok with other encodings - don't have a preference. So feel free to file an issue to switch to a different one.